### PR TITLE
fix: give SSH stdio sessions a chance to close before closing netstack

### PR DESCRIPTION
**Pull request by spikecurtis**
Man, graceful shutdown is hard.  Even after my changes, we were still hitting a graceful shutdown race: https://github.com/coder/coder/runs/18886842123

The problem was that while we attempt a graceful shutdown at the SSH layer by closing the session for writing, we were not giving it a chance to complete before continuing to tear down the stack of closers, including one that closes the netstack, and thus drop the TCP connection before it closes.

### CODE REVIEW

**Comment by spikecurtis** [comment]
Current dependencies on/for this PR:
* main
  * **PR #10815** <a href="https://app.graphite.dev/github/pr/coder/coder/10815?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/coder/coder/10815?utm_source=stack-comment).

**Comment by mafredri** [comment]
Hmm, is my test or my expectation faulty? I tried this PR through the following way:

```sh
while :; do {sleep 10 && pkill -HUP -f /home/coder/src/coder/coder/coder} &; ssh coder.w; sleep 0.1; done
```

While following the agent log, the WireGuard peers keep accumulating:

```
2023-11-21 13:05:53.606 [debu]  net.tailnet.net.wgengine: wgengine: Reconfig: configuring userspace WireGuard config (with 15/15 peers)
```

However, sending the HUP to SSH seems to appropriately tear the connection down and there's no accumulation:

```sh
while :; do {sleep 10 && pkill -HUP ssh} &; ssh coder.w; sleep 0.1; done
```

The difference between the two in agent logs seem to be:

```
2023-11-21 13:04:12.685 [info]  ssh-server: ssh connection complete  remote_addr=[fd7a:115c:a1e0:4088:981c:3dc5:c7d2:eb60]:48513  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  error=EOF
2023-11-21 13:04:12.687 [debu]  ssh-server: copy output done  remote_addr=[fd7a:115c:a1e0:4088:981c:3dc5:c7d2:eb60]:48513  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  id=856ccabe-edda-4a8c-a04c-be1ec941e7a4  bytes=734  error=<nil>
2023-11-21 13:04:12.688 [info]  ssh-server: ssh session returned  remote_addr=[fd7a:115c:a1e0:4088:981c:3dc5:c7d2:eb60]:48513  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  id=856ccabe-edda-4a8c-a04c-be1ec941e7a4  error="signal: killed"
```

```
2023-11-21 13:10:57.054 [info]  ssh-server: ssh connection complete  remote_addr=[fd7a:115c:a1e0:4e3c:beec:a26f:c572:feb3]:43841  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  error="ssh: disconnect, reason 11: disconnected by user"
2023-11-21 13:10:57.056 [debu]  ssh-server: copy output done  remote_addr=[fd7a:115c:a1e0:4e3c:beec:a26f:c572:feb3]:43841  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  id=0cb4cc40-9bbe-4a91-b5d8-dd7e4ba5ebd6  bytes=1534  error=<nil>
2023-11-21 13:10:57.056 [info]  ssh-server: ssh session returned  remote_addr=[fd7a:115c:a1e0:4e3c:beec:a26f:c572:feb3]:43841  local_addr=[fd7a:115c:a1e0:49d6:b259:b7ac:b1b2:48f4]:1  id=0cb4cc40-9bbe-4a91-b5d8-dd7e4ba5ebd6  error="signal: killed"
```

So it seems there are still more situations where we could handle this gracefully? (Given that we handle HUP/INT, I'd expect the shutdown procedure to be proper when it targets the proxy command too.)

**Review by mafredri — APPROVED** [review]

**Review by johnstcn — APPROVED** [review]

File: cli/ssh.go
Comment (mafredri) [reply cli/ssh.go:905]: I figured it was "crossing our fingers", but I'll let Spike chime in there. As for my pov, I think this is the absolute maximum we ever want to hold up the user seeing as the only repercussion is the "old behavior" (wg peer times out after 5 minutes?).

**Comment by spikecurtis** [comment]
> While following the agent log, the WireGuard peers keep accumulating:

That's #7960 and is unrelated to SSH session termination.

-----

It's true that sending SIGHUP to OpenSSH terminates in a slightly different way: OpenSSH sends an explicit `disconnect` message over the SSH connection, and sending SIGHUP to `coder ssh --stdio`  results in the SSH connection getting closed (`EOF`), but the `agentssh` reacts the same way.

**Comment by mafredri** [comment]
> It's true that sending SIGHUP to OpenSSH terminates in a slightly different way: OpenSSH sends an explicit `disconnect` message over the SSH connection, and sending SIGHUP to `coder ssh --stdio` results in the SSH connection getting closed (`EOF`), but the `agentssh` reacts the same way.

I don't feel like the reaction is the same. As observed, one accumulates WireGuard peers whereas the other does not. So there's a difference in how the tailscale connection is torn down. Why can't we exit (wrt tailscale connection) as cleanly by sending HUP to coder as we do when SSH sends the disconnect message? Seems like it should be doable. 🤔

**Comment by spikecurtis** [comment]
> I don't feel like the reaction is the same. As observed, one accumulates WireGuard peers whereas the other does not. So there's a difference in how the tailscale connection is torn down.

That's surprising.  Can you send me the full logs that show Wireguard peers being accumulated in one case but not the other?

**Comment by mafredri** [comment]
@spikecurtis alright, I'll see if I can reproduce tomorrow (logs from earlier are already gone).

File: cli/ssh.go
Comment (spikecurtis) [reply cli/ssh.go:905]: It's long enough to include a couple RTT, even on a slow-but-usable connection, but fast enough that people won't think it's "hung".  Anything in the ballpark of a few to handful of seconds would meet these criteria; it's not exact.

Comment (spikecurtis) [reply cli/ssh.go:905]: Already have the `defer t.Stop()`

**Comment by spikecurtis** [comment]
### Merge activity

* **Nov 22, 4:11 AM**: @spikecurtis merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/10815).

**Comment by mafredri** [comment]
@spikecurtis I tried to reproduce the behavior today but I was unable to, so not sure what/why I observed what I did yesterday.

I'm fairly sure the running agent I was testing against was 7060069034ba977a2c7815cc049603cf7e27ea66, and the client was built from your branch. I can't spot any significant commits that would've changed behavior between tries, so let's forget about this for now.

**Comment by spikecurtis** [comment]
It could just have been a timing issue.  Inactive peers eventually time out and get reaped if there are new peers coming in, making it appear, just by looking at the total number that maybe they were not accumulating in the second case.

**Pull request merged**
